File: 0155-min-stack/0155-min-stack.java
class MinStack {
    Stack<Integer> s;
    Queue<Integer> q;

    public MinStack() {
        s = new Stack();
        q = new PriorityQueue();
    }
    
    public void push(int val) {
        s.push(val);
        q.offer(val);
    }
    
    public void pop() {
        int val = s.pop();
        q.remove(val);
    }
    
    public int top() {
        return s.peek();
    }
    
    public int getMin() {
        return q.peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: 0155-min-stack/0155-min-stack.kt
class MinStack() {
    val s = Stack<Int>();
    val minS = Stack<Int>();

    fun push(`val`: Int) {
        s.push(`val`);    
        if (minS.isEmpty() || minS.peek() >= `val`) minS.push(`val`);
    }

    fun pop() {
        val v = s.pop();
        if (v == minS.peek()) minS.pop();
    }

    fun top(): Int {
        return s.peek();
    }

    fun getMin(): Int {
        return minS.peek();
    }

}

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = MinStack()
 * obj.push(`val`)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */